src/main/kotlin/com/openlattice/indexing/BackgroundExpiredDataDeletionService.kt

```diff
@@ -0,0 +1,250 @@
+                }
+            }
+            ExpirationType.FIRST_WRITE -> {
+                expirationField = Instant.now().minusMillis(expiration.timeToExpiration).toEpochMilli()
+                expirationFieldSQLType = Types.BIGINT
+                comparisonField = "${VERSIONS.name}[1]"
+            }
+            ExpirationType.LAST_WRITE -> {
+                expirationField = OffsetDateTime.ofInstant(Instant.now().minusMillis(expiration.timeToExpiration), ZoneId.systemDefault())
+                expirationFieldSQLType = Types.TIMESTAMP_WITH_TIMEZONE
+                comparisonField = LAST_WRITE.name
+            }
+        }
+        val expiredEntityKeyIds = getExpiredIds(comparisonField, expirationField, expirationFieldSQLType, entitySetId).toSet()
+        if (expiredEntityKeyIds.isNotEmpty()) {
+            val dataTableDeleteStmt = connection.prepareStatement(deleteExpiredDataFromDataTableQuery(entitySetId, comparisonField))
+            dataTableDeleteStmt.setObject(1, expirationField, expirationFieldSQLType)
+            deleteCount = dataTableDeleteStmt.executeUpdate()
+        }
+        return Pair(expiredEntityKeyIds, deleteCount)
+    }
+
+    private fun getExpiredIds(comparisonField: String, expirationField: Any, expirationFieldSQLType: Int, entitySetId: UUID): PostgresIterable<UUID> {
+        return PostgresIterable(
+                Supplier {
+                    val connection = hds.connection
+                    val stmt = connection.prepareStatement(getExpiredIdsQuery(entitySetId, comparisonField))
+                    stmt.setObject(1, expirationField, expirationFieldSQLType)
+                    StatementHolder(connection, stmt, stmt.executeQuery())
+                },
+                Function<ResultSet, UUID> { ResultSetAdapters.id(it) }
+        )
+    }
+
+    private fun deleteExpiredDataFromElasticSearch(entitySetId: UUID, entityTypeId: UUID, expiredEntityKeyIds: Set<UUID>): Boolean {
+        return elasticsearchApi.deleteEntityDataBulk(entitySetId, entityTypeId, expiredEntityKeyIds)
+    }
+
+    private fun deleteExpiredDataFromIdsTable(expiredEntityKeyIds: Set<UUID>): Int {
+        hds.connection.use { conn ->
+            val stmt = conn.createStatement()
+            val expiredIdsAsString = expiredEntityKeyIds.joinToString(prefix = "('", postfix = "')", separator = "', '") { it.toString() }
+            return stmt.executeUpdate(deleteExpiredDataFromIdsTableQuery(expiredIdsAsString))
+        }
+    }
+
+    private fun deleteExpiredDataFromDataTableQuery(entitySetId: UUID, comparisonField: String): String {
+        return "DELETE FROM ${DATA.name} WHERE ${ENTITY_SET_ID.name} = '$entitySetId' AND $comparisonField < ?"
+    }
+
+    private fun getExpiredIdsQuery(entitySetId: UUID, comparisonField: String): String {
+        return "SELECT ${ID.name} FROM ${DATA.name} WHERE ${ENTITY_SET_ID.name} = '$entitySetId' AND $comparisonField < ?"
```

Comment:
I think you'll want `${PARTITION.name}` for the appropriate partitions for this entity set in here as well

Reply:
(for performance)

Reply:
+1 (also in the function above)

Reply:
added this!